### ISSUE TITLE
docker run doesn't pull down latest image if the image exists locally

### ISSUE THREAD

**Issue by mlehner616** [moby/moby#13331]
`docker run` should have an `--pull` or `--update` option to pull down the most up to date version of an image tag, then run. The behavior should match `docker build`.

See #4238

**Comment by thaJeztah** [comment]
I think it would be consistent with `docker build --pull`, on the other hand; the `--pull` flag had to be implemented because for *build* there's no alternative way to deal with this, see https://github.com/docker/docker/issues/4238#issuecomment-54322081. For `docker run`, this would work: `docker pull myimage:tag && docker run myimage:tag`

@unclejack you implemented this for `docker build` at the time, any thoughts?

**Comment by mlehner616** [comment]
I am actually using this work-around and generally agree. However, it comes down to UX and consistency. The very same argument could be made for the `--no-cache` flag where you could just `docker rmi myimage:tag && docker build -t myimage:tag .` While they appear to be effectively doing the same, it's not necessarily clear to the user if docker is performing other optimizations with the `--no-cache` flag set that wouldn't happen using the workaround. A simple flag to pull is a lot cleaner.

**Comment by thaJeztah** [comment]
I agree on the UX perspective. But we need to prevent adding features "just because we can". Not entirely against, so don't give up hope yet. Let's see what others think.

Wrt `docker rmi myimage:tag && docker build -t myimage:tag .`. I don't think that's comparable. Doing that won't work if the old `myimage:tag` is still in use by a container, while `--no- cache` does. But I understand why you're comparing.

**Comment by stevenschlansker** [comment]
IMO this is a good indication that tags should be immutable -- especially once caching mirrors start to come into play, the behavior of changing tags becomes incredibly complex and hard to manage.

Additionally, one of the "big selling points" of Docker is that when I `docker run myapp:v3.1.4`, I get the *exact same image* as anybody else doing so.  Mutable tags breaks this promise.  Anecdotally, we already were bitten by this once when an upstream image of `etcd` overwrote a release tag with a "fixed" version that introduced a regression.

**Comment by thaJeztah** [comment]
@stevenschlansker I agree, but changing tags to be immutable will not be possible without causing a serious BC breaking change.

However, https://github.com/docker/docker/issues/10740 and https://github.com/docker/docker/pull/11109 added support for referring images by digest, which does make this possible.

**Comment by eburnette** [comment]
I'd really like to see this feature too. After rebuilding, I'd like to send a single docker command to remote hosts to start a container, but currently I have to send two: docker pull and docker run. 

I found it surprising that docker didn't check for new versions by default. Currently, if the image doesn't exist, it pulls the newest version but from then on it doesn't check or do a pull. Was it done that way for performance reasons?

**Comment by thaJeztah** [comment]
> I found it surprising that docker didn't check for new versions by default. Currently, if the image doesn't exist, it pulls the newest version but from then on it doesn't check or do a pull. Was it done that way for performance reasons?

I think this was discussed at some point, but (if I recall correctly) there were a number of reasons;

- If I have a locally built image (`foobar:latest`), and want to test it *before* pushing, automatically pulling that image from the registry would overwrite the local image.
- Performance; pulling for *each* run can have quite an performance impact.

**Comment by sielaq** [comment]
+1
@mlehner616 has mentioned
>docker run should have an --pull or --update option

which imply that user have to use this option by purpose (it would **not** be a default option)
and have to deal with any consequences.

Still, this option is very needed,
imagine that we have to deal with hundreds docker slaves
if you prefer "consistence of images" over "speed",
then performance can be solved with local registry.

Sync problem we have solved using `mcollective` or `consul exec` to trigger `docker pull` 
whenever new image is released - but not everyone have/use those tools.

**Comment by eburnette** [comment]
@thaJeztah Yes, I can understand now why it was done this way by default, but it was surprising at first. Docker run might get you the latest image or it might get you a year old image, you can't tell.

**Comment by rbair23** [comment]
I was just bitten by this. From an API perspective, the mistake the Docker team made was adding the shortcut to "run" to auto-pull images. The problem is that pulling on "run" becomes a modal operation. How the run command behaves w.r.t. pulling images has 2 modes: it behaves one way if the image has not been previously pulled and another way if it has. This of course is not naturally visible in the API -- you have to read documentation (or spend time debugging as I did and hit the internet) to learn about this mode in the API.

If the run command never auto-pulled, then the API would no longer be modal and it would be predictable (yes the simple intro example is 2 lines now instead of 1 but everybody knows what is going on and how things work and that is a really big deal for usability).

Or you have to make it so that by default "run" always pulls. I can see both reasons that @thaJeztah gives as good reasons not to want to do this.

**Comment by thaJeztah** [comment]
ping @duglin @tiborvass @unclejack (just some random names, sorry) wdyt? I think this is a reasonable request that brings a friendlier UX and parity with `docker build --pull`.

**Comment by adamkdean** [comment]
Just my 2¢, how exactly is `docker pull image && docker run image` supposed to work with Swarm? I really need a way to pull the new image as part of the run command.

**Comment by cpuguy83** [comment]
@rbair23 Actually the API does _not_ automatically pull, this is strictly a client-side implementation, and the right one IMO.

@adamkdean 
To get around this (and in general is good practice anyway), people should take advantage of the new ability to specify an image via it's content addressable digest.
https://docs.docker.com/reference/commandline/cli/#listing-image-digests
Though the UI (both in Hub and the CLI) could be better here, it's a great feature.

**Comment by rbair23** [comment]
@cpuguy83 I was using API in the general term applying it to the programming interface exposed by the 'docker' command, not the REST API.

**Comment by adamkdean** [comment]
@cpuguy83 I got around this by going though each Swarm node and running `docker -H $host pull <image>` prior to the run, but if I can specify a sha1 digest then that might actually be better. Thanks.

**Comment by tiborvass** [comment]
@thaJeztah I think this is not a feature, it's a bug. With the ongoing trust efforts I hope that this will be resolved.

**Comment by thaJeztah** [comment]
@tiborvass more an "omission"; I think the default behavior *does* make sense (see https://github.com/docker/docker/issues/13331#issuecomment-106361207), but adding a `--pull` option is something that needs to be added.

Now that you completed the "docker daemon" PR, are you perhaps interested in creating a PR? :wink: 

Wrt the reference to "trust"; would that mean that on *every* build, Docker should pull / verify the image?

**Comment by ledjon** [comment]
I'm not really concerned if it is part of default behavior or a --pull parameter, but I really think there should be a way to do this as part of one command.  Perhaps even just spit out a message saying "I am starting this container, but be aware that this is not the latest.  use --pull to get the latest"

Just my thoughts on it.

**Comment by andrerom** [comment]
+1 on this, and +1 on adding this to docker-compose where you just like build don't know about external deps.

**Comment by coderfi** [comment]
`--pull` would be huge!

It is too cumbersome, for every Dockerfile, to:

* figure out what the base image is (awk the FROM line? hard code it in your build script?)

* issue a docker pull for that if it contains the word 'latest' or is omitted (and/or leveraging additional build/CI systems)

* then proceed to build

A generic option like `--pull`, `--pull-always`, `--pull-if-missing`, etc. would be much simpler than trying to do the above in every project.

Yes, you could argue: "well why not just specify the image version".

The problem with that is the case when you are maintaining a series of Dockers which build upon each other.

We'd have to roll in a complex build system to propagate base versions down as they get build up ... and that would mean hacking the downstream Dockerfiles each time... (or dynamically generating them? even messier...)

It is much simpler to just use the 'latest' for every FROM base in our Dockerfiles.

**Comment by andrerom** [comment]
@coderfi This is for `docker run`, support for pull already exists for `docker build` which is what you are looking for as you refer to ´FROM´ lines in DockerFile.

The need to also add this to docker run is most clear for tools like Docker compose being able to very simple being able to add such features for `docker-compose up` which would be huge, and is much needed.

**Comment by bmannix** [comment]
+1 for a `--pull` flag.  This would simplify my workflows.

**Comment by yinchuan** [comment]
+1 for a `--pull` flag

**Comment by coolljt0725** [comment]
I think this is a easy client only implementation.
docker run will call `createContainer` which will pull the image.

**Comment by thaJeztah** [comment]
@coolljt0725 sounds good, are you interested in opening a PR?

**Comment by coolljt0725** [comment]
PR is coming :)

**Comment by thaJeztah** [comment]
Thanks @coolljt0725 !

**Comment by alexw23** [comment]
:+1:

**Comment by Ralf-Te** [comment]
+1
Pls consider increasing priority of the feature request. In my opinion it is more essential as expected at the moment. Let my explain this by the scenario I am currently facing. To carry the issue too far, I will also explain why I think the issue should be flagged as a "bug". :D

I built a docker-based automatically scaling environment, where Amazon ECS controls the docker container placement on the docker hosts dynamically (autoscaling docker containers on auto scaling EC2-Instances). This system is running a production environment!
Because containers are launched dynamically on-demand its critical if old images are used for one or another container. Referring to the issue discussed here we encountered several states of the overall application hard to debug.
It is essential and very cumbersome to manage a cleanup of "n" docker hosts before the deployment of "m" docker containers in such a dynamic environment. 
Especially on platforms like Amazon ECS, etc. where a cloud provider has already built up complex processes and thus is restricting its user to defined patterns the topic, becomes more critical.
For us the issue is limiting us in our possibilities even if automated and also having a high degree of effort to automate/test/monitor/maintain.

As I said, I'd like place a little provocation. :dancer: 
Pls don't feel annoyed, but challenged in your creativity or just don't mind. It's my 2 cent on the issue ;)

Docker should deliver consistency in strictly verifying local state against a defined target state.
From the perspective of a production environment, the --pull option should not be an option it should be the default behavior. This is contrary to the history of docker having its root in local development (Tthis is what i perceived so far from docker and thus why I understand that theirs a fundamental debate on this feature).
However, from my point of view, "dev" should be the mode explicitly called, not "prod" as interpreting --pull as default behavior for "prod" in order to deliver strict consistency in operation. Its what i would expect and thus not have to enable additionally. I'd like to have a diff to "prod" in "dev" to emphasize the difference of "dev" in comparison to "prod" clearly, even if the command line or script is a few lines longer.

**Comment by cpuguy83** [comment]
@ralf-te fist let me just say I am +1 for adding the --pull flag (there is already an PR for this, see above).

In terms of default behavior in prod, I think this is wrong and would encourage people to use content addressable image names (available since 1.8) instead. This would remove the need for using such a flag since you either have the image or you don't since you are referencing a checksum value instead of a friendly name.

**Comment by xujihui1985** [comment]
pull && run on swarm nodes will be very slow and error-prone, suppose we have a 100 node swam cluster, pulling an image will be really time consuming, and there will always the case some of the node failed to download the image. I agree with @stevenschlansker that tag should be immutable, once create, never change. But on the other side when we doing cd in our daily development lots of useless image will be created if we create new tag for every build.  Any good idea? I'm currently using pull && run in for my develop branch and immutable tag for release branch.

**Comment by dominicwilliams** [comment]
Yeah just been stung by this. Couldn't understand why devs were pulling old images despite shared run script appending :latest to the image tag. The behavior between first run and later runs is inconsistent. Will be source of problems bugs imo.

**Comment by cpuguy83** [comment]
@dominicwilliams using `:latest` is most certainly incorrect.
You should use a real pinned version, or the content addressable ID (the later just being a lazy-person's pinned version identifier)

**Comment by jonatanblue** [comment]
:+1:

**Comment by Skilgarriff** [comment]
+1

**Comment by david-antliff-imgtec** [comment]
+1

**Comment by mimmus** [comment]
We deploy our images via systemd which doesn't like long running processes such as docker pull in the ExecStartPre fields and we have a few images, so I don't want to have to have a second service file just to pull the images first. A simple --pull or --update would be great.

**Comment by ugurarpaci** [comment]
--pull +1

**Comment by yaziG** [comment]
+1

**Comment by bean5** [comment]
Unfortunately, even if you use --no-cache, you are affected by this. This lack of pulling makes it really easy for builds to differ between machines, particularly Jenkins slaves, especially when one slave is much older than the other. Automated and non-automated (developer) builds are affected by this.

**Comment by thaJeztah** [comment]
@bean5 if you're having that issue for `docker build`, then there's already `--pull` flag, however if you need builds to be exactly the same, you should use `docker content trust`, or refer to images by *digest*, e.g.: 

```Dockerfile
FROM alpine@sha256:3dcdb92d7432d56604d4545cbd324b14e647b313626d99b889d0626de158f73a
```
or

```
docker run image@sha256:......
```

Alternatively, `docker pull image && docker run ...` should do the trick if you want to always pull.

This issue is still something we want to address, but last PR opened for this, there's was no consensus on design yet; see https://github.com/docker/docker/pull/16609

**Comment by bean5** [comment]
@thaJeztah thanks for that!

**Comment by bean5** [comment]
Running the following on a cron may be appropriate for some uses: `docker images --filter "dangling=false" --format "{{.Repository}}:{{.Tag}}" | xargs -L1 docker pull`.

**Comment by Hubbitus** [comment]
It would be also cool see that flag on restart, so do that:
```
docker restart --pull container_name
```
instead of:
```
docker stop container_name
docker rm -v container_name
docker run --name container_name …
```

The problem there also what I should remember or store somewhere previous run options.

**Comment by thaJeztah** [comment]
@Hubbitus it's unlikely that will be implemented for `docker run`, but you can find similar functionality in `docker service update`, which does exactly that

**Comment by krzyzacy** [comment]
+1

**Comment by sunshineo** [comment]
@krzyzacy  Also click the thumbs up emoji on the OP.

**Comment by krzyzacy** [comment]
@sunshineo I need a :-| emoji

**Comment by qudongfang** [comment]
i really need this feature. +1

**Comment by databill** [comment]
--pull
+1 for this feature

**Comment by mishadev** [comment]
+1

**Comment by hylandm** [comment]
A force pull would be very useful. +1

**Comment by vadyalex** [comment]
I strongly agree that tags should be immutable.

An alternative approach maybe be to have a convention that **"latest"** tag or rather **dev/unstable/snapshot** should be auto-fetched from the Docker Hub if image checksum differs. 

Similar approach has Maven regarding artifacts that have **-SNAPSHOT** in the version number.

**Comment by xussof** [comment]
+1

**Comment by blockjon** [comment]
+1

**Comment by stefan-evinance** [comment]
+1

**Comment by dolobanko** [comment]
I think it's a major issue/feature that should be considered to implement, as it has been reported over 2 years ago.

**Comment by schasse** [comment]
+1

**Comment by bombadiltom** [comment]
+1

**Comment by abdvl** [comment]
+1

**Comment by MarcinMilewski** [comment]
+1

**Comment by centigrade-thomas-becker** [comment]
Please stop the +1 spam

**Comment by kyprifog** [comment]
sigh...

**Comment by montanaflynn** [comment]
The kubernetes equivalent would be `imagePullPolicy` which is something I always put on. 

> The default pull policy is IfNotPresent which causes the Kubelet to skip pulling an image if it already exists. If you would like to always force a pull, you can do one of the following:
set the imagePullPolicy of the container to Always.

**Comment by thaJeztah** [comment]
See the continuation in https://github.com/moby/moby/issues/13331, which has a similar proposal

**Comment by montanaflynn** [comment]
@thaJeztah that is this issue

**Comment by thaJeztah** [comment]
@montanaflynn oops, copy/pasta mistake: updated

**Comment by kwerle** [comment]
Specifically, I am running docker as a child process command.  It's fairly easy to invoke a single command and have it run and attach to its stdin/out and all that good stuff.  It's significantly more complicated to have to try to pull and then run the actual command that I want to run whether the pull succeeds or not.

**Comment by lanoxx** [comment]
How about having a separation similar to git where git distinguishes between local and origin, so you could do something like this:

    docker run foo-image:origin/latest  #pulls latest version from docker hub and runs it
    docker run foo-image:latest         #uses the version tagged as `latest` from the local repository

**Comment by chenliu0831** [comment]
LOL can't believe this issue is still around! that's the NO.1 docker gotchas I think

**Comment by thaJeztah** [comment]
Thanks for your constructive comment @chenliu0831, it's appreciated

**Comment by bean5** [comment]
It is a big gotcha I think. I can't imagine it would be hard to add. Docker
is open source, right? Let's just add it and make a merge request.

**Comment by NicoTexas** [comment]
Rather than an option, I think a convention would be better, like the Maven SNAPSHOT one.
I am thinking of compatibility with Compose for example, but also that I would turn on this option almost everywhere, which at the end would create more risks and break the immutability contract of Docker images.

**Comment by bean5** [comment]
I think if docker-compose had built-in support of dev vs non-dev
environment (and therefore docker-compose.yml vs docker-compose-dev.yml, it
could ask you on the CLI at some point whether you want to enable). Just
one idea of many.

**Comment by purnendu91** [comment]
Is this feature included /developed in the current version?

Or do we still have to 
docker stop container_name
docker rm -v container_name
docker run --name container_name …
?

**Comment by cpuguy83** [comment]
No.

**Comment by Zanadar** [comment]
Gotcha-ed me just today. And LOL there's a 3 year old issue.

**Comment by Zanadar** [comment]
I’ve got a PR open to the CLI that should address this, if anybody wants to review: https://github.com/docker/cli/pull/1498

**Comment by rishiloyola** [comment]
any updates on this issue?

**Comment by Dmitry1987** [comment]
The default behavior of `docker run` right now in Docker `Version 18.03.1-ce` does not pull newer image from the repo :O  is this for real?

**Comment by Dmitry1987** [comment]
it should always check if the tagged image changed in the repository (what if an urgent security fix was patched to the image, and it's expected to be distributed to all infra at the same time?), does it even makes sense that by default Docker does not check she image in the repo? That is SO weird.

**Comment by koenlek** [comment]
There are some good reasons not to have it pull by default (or at least to expose both the options , i.e. to auto-pull or not to the user):
- you often don't want to pull `latest` tags, because it can include breaking changes that you need to check and prepare for, before applying them in your production environment. Note that latest tags are regular tags, there is nothing special about them compared to other tags (afaik). It's just a best practice, but there is no guarantee whatsoever on how a maintainer follow the 'latest' best practice (e.g. will it always point to latest stable image? or might it also point to latest overall image, including pre-releases?).
  - Some popular images in docker-compose setups, like e.g. [webhippie/mariadb](https://hub.docker.com/r/webhippie/mariadb/tags/), manage to *only* have a latest tag *and* make braking changes to them. You should be very careful with accidentally pulling them. It happened to me once that I pulled a newer version of `webhippie/mariadb:latest`, which contained a major mariadb upgrade (according to semantic versioning) with braking changes that broke my setup. You'd better have db backups before that happens to you ;)
- Some maintainers manage to make breaking changes even in the same tags (even if it's not a `latest` tag). I experienced that with the official owncloud images before.

Bottom line: you first want to make sure that you trust a specific maintainer to provide you with a safely auto-pullable tag and/or have proper backup practices in place before you enable auto-pull on `docker run` in many cases. 

Lastly: even if you'd want to change the default, please consider that you might get a whole lot of angry admins  after you after they realize that suddenly the pulling behavior changed and started breaking there setups. Some admins consciously rely on the assumption that new version of a tag won't be pulled automatically.

**Comment by kwerle** [comment]
I don't want it pull by default.  I want the option to make it always check when it's run.

**Comment by Dmitry1987** [comment]
@koenlek  I agree with the last statement that there are many who use a random tag that 'just works' and will be surprised by the sudden change, in docker, **especially if they also have apt/yum auto update turned on** and one morning their docker version on all hosts gets auto-upgraded AND it pulls all images because of the auto-pull feature 🤣  ahahaha

**Comment by Dmitry1987** [comment]
But I still want the auto-pull flag in docker! Very surprised that it doesn't exist.. sounds more like a feature that had to be there at v0.0.2 release, and not discussed during v1.19.X release, that's crazy.

**Comment by cpuguy83** [comment]
Ultimately this is a docker/cli issue.
I recall there was a pr for this awhile ago which got stalled on UX. Might be worth digging through that again.

**Comment by Zanadar** [comment]
*cough* https://github.com/moby/moby/issues/13331#issuecomment-436380634

**Comment by cpuguy83** [comment]
Ah, thanks! Didn't realize this got submitted. Sorry for the slow turn-around.

**Comment by Dmitry1987** [comment]
@Zanadar is there a way to vote on the PR so it gets attention from the maintainers? If there is, you can count on me 👍  +1 , just tell me where to click

**Comment by Zanadar** [comment]
@dmitry1987 feel free to click through on that PR and upvote their

**Comment by Zanadar** [comment]
@mlehner616 #13331 (comment) has been merged and should close this issue (I believe).

**Comment by thaJeztah** [comment]
Ah, yes, https://github.com/docker/cli/pull/1498 has been merged, but missed the cut-off for Docker 19.03, so will ship as part of Docker 19.09 (but you can download nightly builds with that change)

Let me close this one

**Comment by fescobar** [comment]
```
#!/bin/bash

# This is not the best option to pull the latest images but it's a workaround.
# When you try to pull an image with a tag that already exists in your machine, Docker by default doesn't pull the image.
# With this workaround, I'm comparing the Digest (sha256) what I have locally against what I have in my docker hub remotely.
# If there is a difference, then I remove my local image and I will pull the latest from the remote hub.
# Otherwise I will keep using the same image existing in my machine. 

# EXAMPLE:
DOCKER_REGISTRY=docker.io
DOCKER_NAMESPACE=frankescobar
DOCKER_IMAGE=allure-docker-service
DOCKER_TAG=2.13.0

# Use 'DOCKER_ENCODED_AUTH' for private repositories. Use user:pass encoded in base64. 
# Example
# frank:somepass
# ZnJhbms6c29tZXBhc3M=
DOCKER_ENCODED_AUTH=ZnJhbms6c29tZXBhc3M=


DOCKER_API_TOKEN=$(curl -X GET "https://auth.docker.io/token?service=registry.$DOCKER_REGISTRY&scope=repository:$DOCKER_NAMESPACE/$DOCKER_IMAGE:pull" --header "Authorization: Basic $DOCKER_ENCODED_AUTH" | python -c "import sys,json; print json.load(sys.stdin)['access_token']")
REMOTE_TAG_SHA_DIGEST=$(curl -X GET https://registry.hub.docker.com/v2/$DOCKER_NAMESPACE/$DOCKER_IMAGE/manifests/$DOCKER_TAG --header "Authorization: Bearer $DOCKER_API_TOKEN" --header "Accept: application/vnd.docker.distribution.manifest.v2+json" | python -c "import sys,json; print json.load(sys.stdin)['config']['digest']")


LOCAL_TAG_SHA_DIGEST=$(docker inspect --format='{{index .Id}}' $DOCKER_NAMESPACE/$DOCKER_IMAGE:$DOCKER_TAG) || true


if [ -n "$LOCAL_TAG_SHA_DIGEST" ] ; then
    if [ "$LOCAL_TAG_SHA_DIGEST" != "$REMOTE_TAG_SHA_DIGEST" ]; then
        docker image rm -f $DOCKER_NAMESPACE/$DOCKER_IMAGE:$DOCKER_TAG
    fi
fi

docker pull $DOCKER_REGISTRY/$DOCKER_NAMESPACE/$DOCKER_IMAGE:$DOCKER_TAG
```
Requirements: 
- Python
Note:
- It will pull the image even if appears an error.
